pkg/inventory/inventory-client.go

```diff
@@ -27,39 +27,47 @@ type InventoryClient interface {
 	// GetCluster returns the set of previously applied objects as ObjMetadata,
 	// or an error if one occurred. This set of previously applied object references
 	// is stored in the inventory objects living in the cluster.
-	GetClusterObjs(inv *unstructured.Unstructured) ([]object.ObjMetadata, error)
+	GetClusterObjs(inv InventoryInfo) ([]object.ObjMetadata, error)
 	// Merge applies the union of the passed objects with the currently
 	// stored objects in the inventory object. Returns the slice of
 	// objects which are a set diff (objects to be pruned). Otherwise,
 	// returns an error if one happened.
-	Merge(inv *unstructured.Unstructured, objs []object.ObjMetadata) ([]object.ObjMetadata, error)
+	Merge(inv InventoryInfo, objs []object.ObjMetadata) ([]object.ObjMetadata, error)
 	// Replace replaces the set of objects stored in the inventory
 	// object with the passed set of objects, or an error if one occurs.
-	Replace(inv *unstructured.Unstructured, objs []object.ObjMetadata) error
+	Replace(inv InventoryInfo, objs []object.ObjMetadata) error
 	// DeleteInventoryObj deletes the passed inventory object from the APIServer.
-	DeleteInventoryObj(inv *unstructured.Unstructured) error
+	DeleteInventoryObj(inv InventoryInfo) error
 	// SetDryRunStrategy sets the dry run strategy on whether this we actually mutate.
 	SetDryRunStrategy(drs common.DryRunStrategy)
+	// ApplyInventoryNamespace applies the Namespace that the inventory object should be in.
 	ApplyInventoryNamespace(invNamespace *unstructured.Unstructured) error
+	// InvInfoFactoryFunc returns the factory function to create an InventoryInfo from an unstructured.
+	InvInfoFactoryFunc() UnstructuredToInvInfoFunc
```

Comment:
Why is this needed? I think this would no longer be needed if we just remove the `SplitUnstructured` function completely from the apply-as-a-library code. In the new structure, the library should never need to do this.

Reply:
This is needed since the inventory object is split from the resource list. The underlying object of `InventoryInfo` is the inventory object, which is not included in the resource list passed into `Apply` or `Run`. That is my previous understanding. If you mean the inventory object should be part of the resource list, then we don't need the `SplitUnstructured` function or the new function added here.

Reply:
Removed this since the Split function is hidden behind the `provider.InventoryInfo()`.

Reply:
So I think this is an improvement. My concern still is that we are mixing up the features needed in the cobra commands (which isn't really something we expect others to reuse) from apply-as-a-library (which we do want to make it easy for others to use). I think we should consider breaking the Provider interface into two separate interfaces:
 * The Provider interface that will contain the `InventoryClient`, `Factory` and `ToRESTMapper` functions (maybe even remove the `ToRESTMapper` function since that is already available through the Factory).
 * Create a separate `ManifestLoader` (open to suggestions for better names here) that will contain the `ManifestReader` and `InventoryInfo` functions. This interface is probably temporary, but it allows us to reuse the cobra commands without polluting the interfaces for apply-as-a-library with things that aren't needed.

The `Get<Apply/Preview/Destroy>Runner` functions will accept both interfaces, but the `NewApplier` function will only need the Provider interface.
WDYT?

Reply:
Agreed with breaking the provider into separate interfaces. One is for parsing/loading resources and the other is for applying. The parsing interface `ManifestLoader` is only used in the cobra commands. It may not be temporary since parsing is also needed in other client. Later we can also put that into its own library. 

`ManifestLoader` can contain `ManifestReader` and `InventoryInfo`. Or we can change the interface of `ManifestReader`, where the reader returns both `InventoryInfo` and `[]*Unstructured`.

Reply:
Updated the change for splitting the interface. PTAL.